### PR TITLE
V8 master

### PR DESCRIPTION
**Pull request by SimonRichardson**
This brings in the changes from v8 branch into master.

1eb0c22 (upstream/v8) Merge pull request #330 from SimonRichardson/arch-series-identifier-url
fd52ba4 Merge pull request #329 from SimonRichardson/arch-series-identifier-url

### CODE REVIEW

File: url.go [hunk 1]
```diff
@@ -473,14 +525,37 @@ func parseIdentifierURL(url *gourl.URL) (*URL, error) {
...
 	parts := strings.Split(strings.Trim(path, "/"), "/")
-	if len(parts) != 1 {
-		return nil, errors.Errorf(`charm or bundle URL %q malformed, expected "<name>"`, url)
+	if len(parts) == 0 || len(parts) > 3 {
+		return nil, errors.Errorf(`charm or bundle URL %q malformed`, url)
 	}
 
-	r.Name, r.Revision = extractRevision(parts[0])
+	var nameRev string
+	if num := len(parts); num == 3 {
```
Comment (manadart) [inline url.go:533]: Switch?

File: url.go [hunk 2]
```diff
@@ -473,14 +525,37 @@ func parseIdentifierURL(url *gourl.URL) (*URL, error) {
...
+		r.Architecture, nameRev = parts[0], parts[1]
+	} else {
+		nameRev = parts[0]
+	}
+
+	// Mandatory
+	r.Name, r.Revision = extractRevision(nameRev)
 	if err := ValidateName(r.Name); err != nil {
-		return nil, errors.Annotatef(err, "cannot parse URL %q", url)
+		return nil, errors.Annotatef(err, "cannot parse name and/or revision in URL %q", url)
```
Comment (manadart) [inline url.go:544]: I always annotate with what we were doing, say `parsing name and/or revision in URL`. The fact that we couldn't do it is implied by the error.

**Review by manadart — APPROVED** [review]
Looks good. Minor nits.

File: url.go (hunk 2 above)
Comment (SimonRichardson) [reply url.go:544]: Agreed, but this is the convention in this package, so I'll keep it for now.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**